azkaban-web-server/src/main/java/cloudflow/services/FlowServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+package cloudflow.services;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import azkaban.user.User;
+import cloudflow.daos.FlowDao;
+import cloudflow.daos.FlowVersionDao;
+import cloudflow.daos.ProjectDao;
+import cloudflow.error.CloudFlowException;
+import cloudflow.error.CloudFlowNotFoundException;
+import cloudflow.error.CloudFlowValidationException;
+import cloudflow.models.FlowResponse;
+import cloudflow.models.FlowVersion;
+import cloudflow.models.Project;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlowServiceImpl implements FlowService {
+
+  private static final String PROJECT_ID_PARAM = "project_id";
+  private static final String PROJECT_VERSION_PARAM = "project_version";
+  private static final String PROJECT_NAME_PARAM = "project_name";
+  private static final String EXPAND_PARAM = "expand";
+  private static final String VERSIONS_KEY = "versions";
+  private static final Logger log = LoggerFactory.getLogger(ProjectServiceImpl.class);
+  private final FlowDao flowDao;
+  private final FlowVersionDao flowVersionDao;
+  private final ProjectDao projectDao;
+
+  @Inject
+  public FlowServiceImpl(FlowDao flowDao, FlowVersionDao flowVersionDao, ProjectDao projectDao) {
+    this.flowDao = flowDao;
+    this.projectDao = projectDao;
+    this.flowVersionDao = flowVersionDao;
+  }
+
+
+  @Override
+  public FlowResponse getFlow(String flowId, Map<String, String[]> queryParamMap) {
+    requireNonNull(flowId, "Flow id cannot be null");
+    requireNonNull(queryParamMap, "query param map cannot be null");
+
+    Optional<String> expandValue = Optional.empty();
+
+    for (Map.Entry<String, String[]> entry : queryParamMap.entrySet()) {
+      String key = entry.getKey();
+      String[] value = entry.getValue();
+      if (key.equals(EXPAND_PARAM)) {
+        expandValue = Optional.of(extractArgumentValue(key, value));
+
+        /*Currently only expand=versions is supported */
+        if (!expandValue.get().equals(VERSIONS_KEY)) {
+          throw new CloudFlowValidationException("Invalid value for expand: " + expandValue);
+        }
+      }
+    }
+
+    Optional<FlowResponse> flow = flowDao.getFlow(flowId);
```

Comment:
I would favor a model where the DAO layer is not required to return a `FlowResponse` object. 
Instead the data returned by the `FlowDao` along with any further processing and additions can be converted into a single (hopefully immutable) `FlowResponse` object in one-shot (by a method call perhaps).
Currently a single `FlowResponse` object is mutated by multiple different methods which makes the implementation slightly harder to follow and has an impact on future maintainability if the object starts getting more complex. Let me know your thoughts.

Reply:
As discussed offline, created https://jira01.corp.linkedin.com:8443/browse/LIHADOOP-51816 to track this.